django/contrib/gis/gdal/raster/band.py

```diff
@@ -65,5 +64,85 @@ def nodata_value(self):
         Returns the nodata value for this band, or None if it isn't set.
         """
         nodata_exists = c_int()
-        value = capi.get_band_nodata_value(self.ptr, nodata_exists)
+        value = capi.get_band_nodata_value(self._ptr, nodata_exists)
         return value if nodata_exists else None
+
+    @nodata_value.setter
+    def nodata_value(self, value):
+        """
+        Sets the nodata value for this band.
+        """
+        if not isinstance(value, (int, float)):
+            raise ValueError('Nodata value must be numeric.')
+        capi.set_band_nodata_value(self._ptr, value)
+        self.source._flush()
+
+    def datatype(self, as_string=False):
+        """
+        Returns the GDAL Pixel Datatype for this band.
+        """
+        dtype = capi.get_band_datatype(self._ptr)
+        if as_string:
+            dtype = GDAL_PIXEL_TYPES[dtype]
+        return dtype
+
+    def data(self, data=None, offset=None, size=None, as_memoryview=False):
+        """
+        Reads or writes pixel values for this band. Blocks of data can
+        be accessed by specifying the width and height and the offset of
+        the desired block. The same specification can be used to update
+        parts of a raster by providing an array of values.
+
+        Allowed input data types are bytes, memoryview, list, tuple and array.
+        """
+        if not offset:
+            offset = (0, 0)
+        if not size:
+            size = (self.width - offset[0], self.height - offset[1])
+
+        if any((x <= 0 for x in size)):
+            raise ValueError('Offset too big for this raster.')
+
+        if size[0] > self.width or size[1] > self.height:
+            raise ValueError('Size is larger than raster.')
+
+        # Create ctypes type array generator
+        ctypes_array = GDAL_TO_CTYPES[self.datatype()] * (size[0] * size[1])
+
+        if data is None:
+            # Set read mode
+            access_flag = 0
+            # Prepare empty ctypes array
+            data_array = ctypes_array()
+        else:
+            # Set write mode
+            access_flag = 1
+
+            # This conversion is necessary to make the ctypes conversion
+            # working in both Python 2 and 3 (the use of the memoryview
+            # directly fails in python 2).
+            if isinstance(data, memoryview):
+                data = data.tobytes()
```

Comment:
I removed the use of `buffer` to make this compatible with Python 3. I ran into a small problem with the  `from_buffer_copy` ctypes method, in 2.7 this method throws an error when passing a memoryview. As a workaround, I convert the data to bytes first, then it is passed to ctypes as bytes.